### DISCUSSION TITLE
What's the best practice to load default configurations?

### DISCUSSION THREAD

**Discussion by ShizuhaAki** [fyne-io/fyne#4018] [Q&A] [answered]
I'm developing an app that need to have a default configuration builtin. Users also need to be able to customize it.

I'm wondering what's the best way to load that default config, besides calling the functions of the Preferences API one by one?

The config looks like this:

```json
{
	"languages": [
		{
			"name": "C++14",
			"suffix": "cpp",
			"beforeCommand": "g++ {{FULL}} -o {{FILE}}",
			"execCommand": "./{{FILE}}",
			"cleanupCommand": "rm ./{{FILE}}"
		}
	]
}
```

**Comment by Jacalz** [comment]
Using the latest stable, it is likely the best option to either call the methods on by one or save your JSON as a string store that once instead. The new lists API in  https://github.com/fyne-io/fyne/pull/3932 might be useful once we release v2.4.0 as well.

**Comment by ShizuhaAki** [comment]
Another solution that might be useful: When load configuration, first check if we _can_ read from fyne's Preferences API, if so we use it; if not, we fallback to a JSON that's compiled into the binary.

**Reply by andydotxyz** [reply]
The Preference API will always work... in what situation have you found it non-functional?

To have fallbacks you can use the "...WithFallback" methods on Preferences and avoid the need to read any JSON manually.

**Reply by ShizuhaAki** [reply, marked as answer]
Let me rephrase that... When we founnd that nothing has been loaded into Preferences, we fallback to the builtin JSON.

Or I can just use the Fallback-version and integrate the default configuration into it?

**Reply by andydotxyz** [reply]
Indeed - we have coded this in so you don't have to

**Comment by andydotxyz** [comment]
> I'm wondering what's the best way to load that default config, besides calling the functions of the Preferences API one by one?

The benefit of calling them "one by one" is that you can monitor for changes using data binding and the app will only refresh the impacted items - so it's very efficient.
